pkg/certificate/manager.go

```diff
@@ -208,28 +140,75 @@ func (m *Manager) rotate() error {
 		if err != nil {
 			return errors.Wrapf(err, "failed creating server key/cert for service %+v", service)
 		}
-		m.applyTLSSecret(service, keyPair)
+		err = m.applyTLSSecret(service, keyPair)
+		if err != nil {
+			return errors.Wrapf(err, "failed applyting TLS secret %s", service)
+		}
 	}
 
 	return nil
 }
 
 // nextRotationDeadline returns a value for the threshold at which the
 // current certificate should be rotated, 80%+/-10% of the expiration of the
-// certificate.
+// certificate or force rotation in case the certificate chain is faulty
 func (m *Manager) nextRotationDeadline() time.Time {
-	if m.caCert == nil {
-		m.log.Info("Certificates not created, forcing roration")
+	err := m.verifyTLS()
+	if err != nil {
+		// Sprintf is used to prevent stack trace to be printed
+		m.log.Info(fmt.Sprintf("Bad TLS certificate chain, forcing rotation: %v", err))
+		return m.now()
+	}
+
+	caCert, err := m.getFirstCACertFromCABundle()
+	if err != nil {
+		m.log.Info("Failed reading first CA cert from CABundle, forcing rotation", "err", err)
 		return m.now()
 	}
-	notAfter := m.caCert.NotAfter
-	totalDuration := float64(notAfter.Sub(m.caCert.NotBefore))
-	deadline := m.caCert.NotBefore.Add(jitteryDuration(totalDuration))
+
+	return m.nextRotationDeadlineForCert(caCert)
+}
+
+// nextRotationDeadlineForCert returns a value for the threshold at which the
+// current certificate should be rotated, 80%+/-10% of the expiration of the
+// certificate
+func (m *Manager) nextRotationDeadlineForCert(certificate *x509.Certificate) time.Time {
+	notAfter := certificate.NotAfter
+	totalDuration := float64(notAfter.Sub(certificate.NotBefore))
+	deadline := certificate.NotBefore.Add(jitteryDuration(totalDuration))
 
 	m.log.Info(fmt.Sprintf("Certificate expiration is %v, rotation deadline is %v", notAfter, deadline))
 	return deadline
 }
 
+func (m *Manager) nextElapsedToRotate() time.Duration {
+	deadline := m.nextRotationDeadline()
+	now := m.now()
```

Comment:
suggestion: since it takes a little time to rotate, shouldn't we try to rotate some time in advance **before** the cert expires?
Could be easily changed by setting the `now` to x minutes before now, if i'm not mistaken.
what do you think?

Reply:
That's already done, this rotates 10/30 % of time before expiration with some random jitter.